### DISCUSSION TITLE
Windows 10 fails to rekey AES256-GCM ikev2 session when initiated by server

### DISCUSSION THREAD

**Discussion by dgarlans** [strongswan/strongswan#1934] [Q&A]
So I've been doing a lot of testing to stamp out this issue, and i've finally isolated it down as best I can. It's happened on multiple different pfsense deployments with multiple endpoints, and I'm able to reproduce it reliably as well.

Basically, if the firewall initiates the ikev2 p1, it fails, and the session stays open for a while until the retransmission attempts time out, and the tunnel is closed. (Windows then proceeds to take 15 more minutes to realize the tunnel is dead before it shows disconnection, but that's a separate issue)

The logs look like this:

```
Oct  3 20:36:53 | charon | 79083 | 14[IKE] <con-mobile\|47> rekeying IKE_SA failed, peer not responding
Oct  3 20:36:53 | charon | 79083 | 14[IKE] <con-mobile\|47> giving up after 5 retransmits
(snipped a couple more identical retransmits)
Oct  3 20:34:12 | charon | 79083 | 07[NET] <con-mobile\|47> sending packet: from 66.x[4500] to 38.x[4818] (297 bytes)
Oct  3 20:34:12 | charon | 79083 | 07[IKE] <con-mobile\|47> retransmit 1 of request with message ID 0
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> CREATE_CHILD_SA response with message ID 0 processing failed
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> integrity check failed
Oct  3 20:34:08 | charon | 79083 | 15[ENC] <con-mobile\|47> could not decrypt payloads
Oct  3 20:34:08 | charon | 79083 | 15[ENC] <con-mobile\|47> verifying encrypted payload integrity failed
Oct  3 20:34:08 | charon | 79083 | 15[NET] <con-mobile\|47> received packet: from 38.x[4818] to 66.x[4500] (264 bytes)
Oct  3 20:34:08 | charon | 79083 | 15[NET] <con-mobile\|47> sending packet: from 66.x[4500] to 38.x[4818] (297 bytes)
Oct  3 20:34:08 | charon | 79083 | 15[ENC] <con-mobile\|47> generating CREATE_CHILD_SA request 0 [ SA No KE ]
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> IKE_SA con-mobile[48] state change: CREATED => CONNECTING
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> initiating IKE_SA con-mobile[48] to 38.x
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> IKE_SA con-mobile[47] state change: ESTABLISHED => REKEYING
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47>   activating IKE_REKEY task
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> activating new tasks
Oct  3 20:34:08 | charon | 79083 | 15[IKE] <con-mobile\|47> queueing IKE_REKEY task

```

The p2's rekey okay.

I've been able to work around the issue by setting the rekey time up to 16 hours, which means that the windows 8 hour automatic rekey seems to occur, which works successfully, meaning the session stays up just fine.

This always occurs when the P1 encryption is set to to AES-256 GCM (128bit). The P1 Transform can be SHA256 or SHA384, and the P1 DH group can be 14, 20, or 24. I'm using machine certificates for authentication, either regular RSA ones for 14/24, or an ECDSA-384 one for 20, but either way the problem was the same. 

If I set the encryption back to AES-256 instead, it works perfectly every time. The P2 settings don't seem to be relevant, but i've also tried AES256, AES256-GCM, and several different PFS groups as well.

Does anyone have any suggestions on what might be going on? This issue has been happening across the board across multiple different systems with different versions of pfsense and different version of windows 10, with different endpoints on different ISP's and so on. I'm in the process of testing on windows 11 now too.

I've looked around at previous threads and it seems like a lot of previous issues were due to modp2048, and they also seemed to manifest as windows rekeys failing but server rekeys being successful, whereas I'm seeing the exact opposite.

**Comment by dgarlans** [comment]
Same behavior occurs on a Windows 11 device too

**Reply by MathewCNichols** [reply]
Confirming the same problem using strongSwan on embedded host with Windows 11 clients. Pinned it down to GCM on Phase 1. I'll also band-aid it by not using GCM on Phase 1 or extending ikelifetime to 8 1/2 hours.

**Comment by tobiasbrunner** [comment]
Just to clarify, AES-GCM works in general? For instance, DPDs or CHILD_SA rekeying work fine, even an IKE_SA rekeying initiated by the client completes successfully (and messages are exchanged successfully afterwards, e.g. to delete the old IKE_SA or to rekey CHILD_SAs on the new IKE_SA)? But if the server initiates the IKE_SA rekeying, the response can't be decrypted afterwards?

If so, that sounds like a bug in Windows (also see e.g. [this report](https://wiki.strongswan.org/issues/3400#note-4) from three years ago that mentioned problems with AES-GCM as well). Maybe they switch algorithms or keys in a weird way and then send the response incorrectly encrypted.

**Reply by dgarlans** [reply]
Your clarification is correct. I did find that report from three years ago and unfortunately that's the one i was seeing the exact opposite behavior of - they could manually rekey from strongswan, but windows couldn't initiate a rekey itself.

would it be worth cranking up the logging level to try and see if it's possible to tell what windows is sending back? i did wireshark it before but as basically everything was encrypted it didn't help very much.

**Reply by tobiasbrunner** [reply]
> I did find that report from three years ago and unfortunately that's the one i was seeing the exact opposite behavior of - they could manually rekey from strongswan, but windows couldn't initiate a rekey itself.

Yes, the original report was about client-initiated rekeyings. But where do you see that in relation to AES-GCM? The user just said: "Windows's AES-GCM implementation also appeared to be buggy and sent undecryptable packets".  I don't see anything about the exact messages/exchanges this occurred in.

> would it be worth cranking up the logging level to try and see if it's possible to tell what windows is sending back? i did wireshark it before but as basically everything was encrypted it didn't help very much.

Maybe but it will be tricky. For instance, level 3 in _enc_ will log details about the decryption attempt, e.g. IV and ICV values, but that will not immediately tell us anything. With the IKE keys (level 4 in _ike_ logs those), and that data or maybe even a corresponding pcap file we could do some experiments (the latter could be analyzed in Wireshark where you can also configure the keys, maybe using different algorithms to see if e.g. the client does something weird like switching to AES-CBC).

**Comment by tribals** [comment]
It would be interesting to know how you're managed to setup AES GCM in general. Tried many times with `Set-VPNConnectionIPsecConfiguration` but can't get anything beyond AES CBC. Currently it is as follows:

```PowerShell
(Get-vpnconnection -name $conn).IPsecCustomPolicy

AuthenticationTransformConstants : None
CipherTransformConstants         : AES256
DHGroup                          : ECP384
IntegrityCheckMethod             : SHA384
PfsGroup                         : ECP384
EncryptionMethod                 : AES256
```

**Comment by dgarlans** [comment]
I have to check but I think it's just GCMAES256 as the cypher type.
